lib/Core/BuildEngine.cpp

```diff
@@ -1060,14 +1074,48 @@ class BuildEngineImpl {
     return currentTimestamp;
   }
 
+  KeyID getKeyID(const KeyType& key) {
+    // Delegate if we have a database.
+    if (db) {
+      return db->getKeyID(key);
+    }
+
+    // Otherwise use our builtin key table.
+    {
+      std::lock_guard<std::mutex> guard(keyTableMutex);
+      auto it = keyTable.insert(std::make_pair(key, false)).first;
+      return (KeyID)(uintptr_t)it->getKey().data();
+    }
+  }
+
   RuleInfo& getRuleInfoForKey(const KeyType& key) {
+    auto keyID = getKeyID(key);
+    
     // Check if we have already found the rule.
-    auto it = ruleInfos.find(key);
+    auto it = ruleInfos.find(keyID);
     if (it != ruleInfos.end())
       return it->second;
 
     // Otherwise, request it from the delegate and add it.
-    return addRule(delegate.lookupRule(key));
+    return addRule(keyID, delegate.lookupRule(key));
+  }
+
+  RuleInfo& getRuleInfoForKey(KeyID keyID) {
+    // Check if we have already found the rule.
+    auto it = ruleInfos.find(keyID);
+    if (it != ruleInfos.end())
+      return it->second;
+
+    // Otherwise, we need to resolve the full key so we can request it from the
+    // delegate.
+    KeyType key;
+    if (db) {
+      key = db->getKeyForID(keyID);
+    } else {
+      key = llvm::StringMapEntry<bool>::GetStringMapEntryFromKeyData(
```

Comment:
Good question, but no... we don't because once the string map entry is allocated it never moves -- so if we have a key pointer, we don't need to worry about it moving.

Deserves a comment though...